### PR TITLE
Documentation Refactoring

### PR DESCRIPTION
**Pull request by limarta**
This PR restructures the documentation to somewhat resemble the "Documentation System" described [here](https://documentation.divio.com). Sections are organized as follows (see `pages.jl`) :
- Home
- Getting Started: This section is for brief and concise examples that a user is likely to be familiar with e.g. linear regression.
- Tutorials: In-depth tutorials that highlight Gen. We further divide the sections into Basic and Advanced.
- How-to Guides: This section lists snippets on "how to do x." These are not tutorials, so no introductory setup is expected.
- API
  - Modeling Language API
  - Standard Inference API
- Explanations and Internals: This section will serve as conceptual material relevant to the design and internals of Gen. Unlike the tutorials and How-to sections, this section offers theory and discussion about Gen for advanced users and is not strictly necessary to use Gen in applications.

Other additions:
- The tutorials now use Documenter.jl's REPL feature to run Julia code. This feature allows tutorial writers to integrate code directly in the markdown files without having to write hardcoded outputs.
- Ref links that were previously broken are now fixed.

### CODE REVIEW

File: docs/README [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Website Docs
```
Comment (sritchie) [inline docs/README]: Change this to `docs/README.md` with the markdown extension, please

File: docs/build_docs_locally.sh [hunk 2]
```diff
@@ -1,6 +0,0 @@
-#!/bin/sh
-
```
Comment (sritchie) [inline docs/build_docs_locally.sh:2]: add a note about this to the docs README, how to invoke it

File: docs/make.jl [hunk 3]
```diff
@@ -1,37 +1,19 @@
...
-            "Particle Filtering" => "ref/pf.md",
-            "Variational Inference" => "ref/vi.md",
-            "Learning Generative Functions" => "ref/learning.md"
-        ],
-        "Internals" => [
-            "Optimizing Trainable Parameters" => "ref/internals/parameter_optimization.md",
-            "Modeling Language Implementation" => "ref/internals/language_implementation.md"
-         ]
-    ]
+    # linkcheck = true,
```
Comment (sritchie) [inline docs/make.jl]: can you remove the commented lines?

File: docs/src/assets/header.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      return nav
+    })())
+    return container
+  })())
+
+  const documenterTarget = document.querySelector('#documenter');
+
+  documenterTarget.parentNode.insertBefore(header, documenterTarget);
+
+  // Edit broken links in the examples, see issue #70
```
Comment (sritchie) [inline docs/src/assets/header.js:85]: where did this come from? Is this a reference to one of our issues? If not, add an attribution for where this came from

File: docs/src/misc/references.md
Comment (sritchie) [inline docs/src/misc/references.md:1]: should you delete these empty files?

File: docs/src/misc/contributing.md
Comment (sritchie) [inline docs/src/misc/contributing.md:1]: delete

File: docs/src/misc/changelog.md
Comment (sritchie) [inline docs/src/misc/changelog.md:1]: delete

File: docs/src/tutorials/languages/static_dsl.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Reduce Runtime with the Static Modeling Language
+
+!!! note
```
Comment (sritchie) [inline docs/src/tutorials/languages/static_dsl.md]: looks like this needs some attention

File: docs/src/tutorials/vi.md [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
 In Gen, variational families are represented as generative functions, and variational inference typically involves optimizing the trainable parameters of generative functions.
 
-## Black box variational inference
-There are two procedures in the inference library for performing black box variational inference.
-Each of these procedures can also train the model using stochastic gradient descent, as in a variational autoencoder.
-```@docs
-black_box_vi!
-black_box_vimco!
+Let's create a generative model representing a joint set of coins.
+
```
Comment (sritchie) [inline docs/src/tutorials/vi.md]: needs some attention, yeah?

**Review by sritchie — CHANGES_REQUESTED** [review]

File: docs/src/tutorials/combinators.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+<div style="text-align:center">
+    <img src="../../images/recurse_combinator.png" alt="schematic of recurse combinatokr" width="70%"/>
+</div>
+```
+
+TODO: document me
+
+```@raw html
+<div style="text-align:center">
+    <img src="../../images/recurse_combinator.png" alt="schematic of recurse combinatokr" width="70%"/>
```
Comment (sritchie) [inline docs/src/tutorials/combinators.md]: typo

File: docs/build_docs_locally.sh (hunk 2 above)
Comment (limarta) [reply docs/build_docs_locally.sh:2]: Moved to README

File: docs/src/tutorials/languages/static_dsl.md (hunk 5 above)
Comment (limarta) [reply docs/src/tutorials/languages/static_dsl.md]: Removed

File: docs/src/tutorials/vi.md (hunk 6 above)
Comment (limarta) [reply docs/src/tutorials/vi.md]: Removed

**Comment by limarta** [comment]
Note that the changes here don't include all of the [tutorials here](https://www.gen.dev/tutorials/). Later tutorial pushes can help convert the remaining tutorials into the required format for the new docs.

File: docs/pages.jl [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        "Modeling Library" => [
+        "Generative Functions" => "api/model/gfi.md",
+        "Probability Distributions" => "api/model/distributions.md",
+        "Built-in Modeling Languages" => "api/model/modeling.md",
+        "Combinators" => "api/model/combinators.md",
+        "Choice Maps" => "api/model/choice_maps.md",
+        "Selections" => "api/model/selections.md",
+        "Optimizing Trainable Parameters" => "api/model/parameter_optimization.md",
+        "Trace Translators" => "api/model/trace_translators.md",
+        "Differential Programming" => "api/model/differential_programming.md"
```
Comment (ztangent) [inline docs/pages.jl]: This should be "Differentiable programming"! In both the label and the filename.

**Comment by ztangent** [comment]
Thanks for this! I'll take a look when I have more time. Is there a link to a rendered version of the new docs, or will I have to build locally?

Also, flagging that if we merge #520 soon, this may require us to update the docs again. #520 should probably a breaking change to Gen.jl though (i.e. it should be released in Gen.jl 0.5.0) --- so my suggestion would be for us to finish this PR, and release the new docs together with Gen.jl 0.4.7, which will be a minor bugfix release with a few new features (e.g. #521, #524).

**Comment by limarta** [comment]
@ztangent You will have to build locally. We should consider adding a dev doc site as well.

**Comment by limarta** [comment]
@ztangent It seems you have to build locally. Not sure why the dev site is not updating.

File: docs/.gitignore [hunk 9]
```diff
@@ -1 +0,0 @@
-Manifest.toml
```
Comment (sritchie) [inline docs/.gitignore:1]: why is this deleted?

File: Project.toml [hunk 10]
```diff
@@ -34,4 +34,4 @@ julia = "1.6"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [targets]
-test = ["Test"]
+test = ["Test"]
```
Comment (sritchie) [inline Project.toml:37]: this is the sort of whitespace change it's nice to not include in a code review...

**Review by sritchie — APPROVED** [review]
LGTM!

**Comment by sritchie** [comment]
NIce work @limarta !!
